### DISCUSSION TITLE
Using KCP-like API server to re-use controller pattern

### DISCUSSION THREAD

**Discussion by invidian** [cosi-spec/enhancements#3] [Ideas]
Not sure if that's the right place to discuss, please redirect if we have a better one.

We talked about possibility of using [KCP](https://github.com/kcp-dev/kcp)-like API server as a runtime, so then configuration management of SO can be done using tooling familiar for people like controller-runtime, kube-builder, Helm etc. 

It seems that concept of [extension-api-server](https://kubernetes.io/docs/tasks/extend-kubernetes/setup-extension-api-server/) is also worth mentioning for that. Our imaginary extension API server could then read e.g. `/proc` on the fly and report resources to `kubectl`. IIRC extension API servers must deal with the storage on their own.

Example of extension API server: https://github.com/kubernetes/sample-apiserver

If we use that approach, we should also consider what roles `controller-manager` plays and if we are going to miss some functionality from it, e.g. garbage collection? 

EDIT: I implemented some quick PoC, available here: https://github.com/invidian/koss.

**Comment by andrewrynhard** [comment]
Thanks @invidian . Very interesting idea on using the extension-api-server!

**Comment by invidian** [comment]
Added PoC link to the OP.
